content/learning-paths/smartphones-and-mobile/using-neon-intrinsics-to-optimize-unity-on-android/5-the-optimizations.md

```diff
@@ -0,0 +1,275 @@
+
+```
+Line 741: results = vqtbl1_u8(_result of ORs_, tblindex1)
+```
+Table lookup function that selects elements from an array based on the indices provided. The result of the OR operations will be treated as an array of 8-bit values. The values from _tblindex1_ (0, 4, 8 and 12) ensure that we select the most significant bytes from each u32 OR result. So 4 character-wall comparisons are being merged into one 128-bit vector along with 4 dummy (because of the _out of range_ values in tblindex1) that will be replaced later with the 64-bit value of the next 4 wall comparisons (from _wmvn_u8_).
+
+```
+Line 751: vqtbx1_u8(results, …)
+```
+Table lookup function except that when an index is out of range, it leaves the existing data alone. This has the effect of selecting 4 bytes (using indices from _tblindex2_) from the results of the last 4 comparisons and combines it with the previous 4 results. _results_ will now contain the results of 8 wall-character comparisons.
+
+```
+Line 751: results = vmvn_u8(...)
+```
+Bitwise NOT operation. This negates each of the 8 character-wall comparisons. It is effectively the _!_ (NOT) in our intersect function; except that it is working on 8 results instead of 1.
+
+```
+Line 755: *(Unity.Burst.Intrinsics.v64*)(collisions + (c * numWalls + w - 4)) = results;
+```
+This is a vectorized form of writing out the 8 results to memory all at once. At time of writing, Burst doesn’t include the _VSTR_ Neon intrinsic.
+
+### Character-character collision detection
+This is left as an exercise for the reader. You may wish to explore the _NeonRadiusObjCollisionDetectionUnrolled_ function starting on line 561 of _CollisionCalculationScript.cs_.
+
+## Additional notes
+The sample project increases the character count over time. Because of this it was better to profile using later frames. In early frames there aren’t enough characters so the overhead of vectorization counters any gains.
+
+An early implementation used Burst Jobs. The jobs got in the way of performance timing because they are not guaranteed to execute immediately.
+
+[NoAlias] had a big impact on the performance gains (without it some optimizations produced little to no gain).
+
+The job implementations have been left in the code for you to experiment with. Leveraging jobs in your own projects can lead to important performance gains when used appropriately.
+
+The sample code displays some timing data on screen. It used _System.Diagnostics.Stopwatch_ to get accurate timings.
+
+It may sound obvious but don’t use the optimizations when they are slower! For the collision sample, the best gain is when there are lots of characters but early on, there aren’t that many. So when there aren’t many characters in the scene, the code could switch to whichever is the fastest version.
```

Comment:
Last potential point, or perhaps it wants to be further up as it's so fundamental - for Burst, but especially for Neon, a lot of it is about the shape of the data. Structure of arrays, not array of structures; putting and storing data in manners that are efficient to grab, process and store as a vector. Know what gets used with what and where and what needs to be next to each other. Simple types.

Reply:
added.
done